File: backend/app/schemas/ask.py
"""Pydantic schemas for the /ask natural-language stats endpoint."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class AskRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)


class AskResponse(BaseModel):
    question: str
    sql: str
    reasoning: str
    columns: list[str]
    rows: list[list[Any]]
    row_count: int
    truncated: bool
    model: str
    llm_latency_ms: int
    db_latency_ms: int
    cache_read_tokens: int
    cached: bool = False
    # Slug of the stored snapshot — the shareable /ask/a/{answer_id} link.
    answer_id: str | None = None


class AskAnswerResponse(BaseModel):
    """A stored answer snapshot, fetched by slug (public) or via history."""

    slug: str
    question: str
    sql: str
    reasoning: str
    columns: list[str]
    rows: list[list[Any]]
    truncated: bool
    model: str
    created_at: datetime

    model_config = {"from_attributes": True}


class AskHistoryItem(BaseModel):
    """Compact history row — question + link, no payload."""

    slug: str
    question: str
    created_at: datetime

    model_config = {"from_attributes": True}

File: web/src/lib/api/ask.ts
/** Natural-language stats: ask, shareable answers, per-user history. */

import { apiFetch } from './client';

export interface AskResponse {
  question: string;
  sql: string;
  reasoning: string;
  columns: string[];
  rows: unknown[][];
  row_count: number;
  truncated: boolean;
  model: string;
  llm_latency_ms: number;
  db_latency_ms: number;
  cache_read_tokens: number;
  cached: boolean;
  answer_id: string | null;
}

export interface AskAnswerResponse {
  slug: string;
  question: string;
  sql: string;
  reasoning: string;
  columns: string[];
  rows: unknown[][];
  truncated: boolean;
  model: string;
  created_at: string;
}

export interface AskHistoryItem {
  slug: string;
  question: string;
  created_at: string;
}

export const askApi = {
  ask(question: string): Promise<AskResponse> {
    // Auth is optional here — send the token when we have one so the ask
    // lands in the user's history; anonymous asks still work.
    return apiFetch('/api/v1/ask', {
      method: 'POST',
      json: { question }
    });
  },
  getAnswer(slug: string): Promise<AskAnswerResponse> {
    return apiFetch(`/api/v1/ask/answers/${slug}`, { skipAuth: true });
  },
  history(): Promise<AskHistoryItem[]> {
    return apiFetch('/api/v1/ask/history');
  }
};
